{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c51db982",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import modules\n",
    "import pysam\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import math\n",
    "import itertools\n",
    "from dataclasses import dataclass\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "063ed1b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def cigar_compute(read):\n",
    "        # Input an primary alignment\n",
    "        # Output \n",
    "        eq , X, I, D, align_qual, M , S,H = 0,0,0,0, 0, 0 , 0,0\n",
    "        for op, length in read.cigartuples:\n",
    "                if op == 7:\n",
    "                        eq += length\n",
    "                elif op == 8:\n",
    "                        X += length\n",
    "                elif op == 1:\n",
    "                        I += length\n",
    "                elif op == 2:\n",
    "                        D += length\n",
    "                elif op == 0:\n",
    "                        M += length\n",
    "                elif op == 4:\n",
    "                        S += length\n",
    "                elif op == 5:\n",
    "                        H += length\n",
    "                else:\n",
    "                        continue\n",
    "        qleng = M+ I + S + eq + X\n",
    "        align_qual = round(float(eq)/(eq+X + I + D),2)\n",
    "        return align_qual, qleng, S, H\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45166b36",
   "metadata": {},
   "source": [
    "# This part contains some plots of the original data. From these observations, we will decide to exclude or include some reads."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9090557f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam\n",
    "\n",
    "file = pysam.AlignmentFile(\"new_map_sorted.bam\",\"rb\")\n",
    "read_dict = {}\n",
    "for read in file.fetch():\n",
    "    if read.is_unmapped or read.is_supplementary or read.is_secondary:\n",
    "        continue\n",
    "    else:\n",
    "        # It is primary alignment\n",
    "        # Do analyze here\n",
    "        CIGAR_qual, qleng, Soft,Hard = cigar_compute(read)\n",
    "        bacteria = read.reference_name[0:2]\n",
    "        if bacteria not in read_dict:\n",
    "            read_dict[bacteria] = 1\n",
    "        else:\n",
    "            read_dict[bacteria] += 1\n",
    "\n",
    "xpoint = []\n",
    "ypoint = []\n",
    "for bacteria in read_dict:\n",
    "    xpoint += [bacteria]\n",
    "    ypoint += [read_dict[bacteria]]\n",
    "    print('The {} has {} alignments.'.format(bacteria , read_dict[bacteria]))\n",
    "\n",
    "plt.title('Distribution of bacteria: primary alignment')\n",
    "plt.plot(xpoint,ypoint,'o')\n",
    "plt.ylabel('Number of alignments')\n",
    "plt.xlabel('Bacteria names')\n",
    "plt.savefig('Bacteria_distribution.png')\n",
    "#plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c5d1107",
   "metadata": {},
   "outputs": [],
   "source": [
    "count = 1\n",
    "qleng_distribution = {}\n",
    "len_min = float('inf')\n",
    "len_max = float('-inf')\n",
    "for read in file.fetch():\n",
    "    if count <3900000:\n",
    "        if read.is_unmapped or read.is_supplementary or read.is_secondary:\n",
    "            continue\n",
    "        else:\n",
    "            # It is primary alignment\n",
    "            # Do analyze here\n",
    "            qleng = cigar_compute(read)[1]\n",
    "            if qleng < len_min:\n",
    "                len_min = qleng\n",
    "            if qleng > len_max:\n",
    "                len_max = qleng\n",
    "            bacteria = read.reference_name[0:2]\n",
    "            if bacteria not in qleng_distribution:\n",
    "                qleng_distribution[bacteria] = [qleng]\n",
    "            else:\n",
    "                qleng_distribution[bacteria] += [qleng]\n",
    "        if count%10000 == 0:\n",
    "            print('Pass 100000K')\n",
    "    else:\n",
    "        break\n",
    "    count+=1\n",
    "\n",
    "plt.savefig('Bacteria_distribution.png')\n",
    "#print('Done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70c277af",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_median(sorted_list):\n",
    "    indices = []\n",
    "\n",
    "    list_size = len(sorted_list)\n",
    "    median = 0\n",
    "\n",
    "    if list_size % 2 == 0:\n",
    "        indices.append(int(list_size / 2.0) - 1)  # -1 because index starts from 0\n",
    "        indices.append(int(list_size / 2.0))\n",
    "\n",
    "        median = (sorted_list[indices[0]] + sorted_list[indices[1]]) / 2.0\n",
    "        pass\n",
    "    else:\n",
    "        indices.append(int(list_size / 2.0))\n",
    "\n",
    "        median = sorted_list[indices[0]]\n",
    "        pass\n",
    "\n",
    "    return median, indices\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d5b05c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "bac_key = ['lf','lm','cn','ef','ec','pa','bs','sc','sa','se']\n",
    "Q1_plot = []\n",
    "med_plot = []\n",
    "Q3_plot = []\n",
    "for bac in bac_key:\n",
    "    samples = qleng_distribution[bac]\n",
    "    samples.sort()\n",
    "    median, median_indices = find_median(samples)\n",
    "    Q1, Q1_indices = find_median(samples[:median_indices[0]])\n",
    "    Q2, Q2_indices = find_median(samples[median_indices[-1] + 1:])\n",
    "\n",
    "    quartiles = [Q1, median, Q2]\n",
    "    Q1_plot += [Q1]\n",
    "    med_plot += [median]\n",
    "    Q3_plot += [Q2]\n",
    "    print(\"For bacteria {}, the quartiles are (Q1, median, Q3): {}\".format(bac,quartiles))\n",
    "    \n",
    "plt.xlabel('Bacteria')\n",
    "plt.ylabel('Alignment length')\n",
    "plt.title('Quartiles')\n",
    "plt.plot(bac_key,Q1_plot,'x', label = 'Q1')\n",
    "plt.plot(bac_key,med_plot,'o', label = 'Med')\n",
    "plt.plot(bac_key,Q3_plot,'x', label = 'Q3')\n",
    "plt.legend()\n",
    "plt.savefig('Alignment_quartiles.png')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2689eec8",
   "metadata": {},
   "outputs": [],
   "source": [
    "bac_key = ['lf','lm','cn','ef','ec','pa','bs','sc','sa','se']\n",
    "fig, axi = plt.subplots(nrows=10, ncols=1, figsize = (24,14), tight_layout = True, sharey = True)\n",
    "\n",
    "for i in range(len(bac_key)):\n",
    "    # Add subplots\n",
    "    axi[i].hist(x=qleng_distribution[bac_key[i]], bins=np.linspace(len_min, len_max,50), color='r',alpha=0.9, rwidth=0.85)\n",
    "    # Show texts\n",
    "    axi[i].text(0.5, 0.5, bac_key[i], horizontalalignment='center', verticalalignment='center', transform=axi[i].transAxes)\n",
    "    #axi[1].text(0.5, 0.5, 'lm', horizontalalignment='center', verticalalignment='center', transform=axi[1].transAxes)\n",
    "\n",
    "plt.ylabel('Number of alignments')\n",
    "plt.xlabel('Length of alignments')\n",
    "plt.savefig('Bacteria_distribution_alignment_length.png')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c509c619",
   "metadata": {},
   "outputs": [],
   "source": [
    "n, bins, patches = plt.hist(x=qleng_distribution['lm'], bins=np.linspace(len_min, len_max,100), color='r',\n",
    "                            alpha=0.9, rwidth=0.85)\n",
    "plt.grid(axis='y', alpha=0.75)\n",
    "plt.xlabel('Alignment Length')\n",
    "plt.ylabel('Number of alignments')\n",
    "plt.title('Histogram of {}'.format('lm'))\n",
    "# Set a clean upper y-axis limit.\n",
    "plt.ylim(ymax=500000)\n",
    "plt.savefig('Bacteria_distribution_alignment_length.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dcbfc259",
   "metadata": {},
   "outputs": [],
   "source": [
    "b = [i/100.0 for i in range(90,100)]\n",
    "plt.hist(x=align_qual_dist['lf'], bins=b, color='r',\n",
    "                          alpha=0.9, rwidth=0.85)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "731b4205",
   "metadata": {},
   "outputs": [],
   "source": [
    "bac_key = ['lf','lm','cn','ef','ec','pa','bs','sc','sa','se']\n",
    "fig, axi = plt.subplots(nrows=10, ncols=1, figsize = (24,20), sharey = True)\n",
    "b = [i/100.0 for i in range(90,100)]\n",
    "for i in range(len(bac_key)):\n",
    "    # Add subplots\n",
    "    axi[i].hist(x=align_qual_dist[bac_key[i]], bins=b, color='r',alpha=0.9, rwidth=0.85)\n",
    "    # Show texts\n",
    "    axi[i].text(0.2, 0.5, bac_key[i], horizontalalignment='center', verticalalignment='center', transform=axi[i].transAxes)\n",
    "    \n",
    "plt.ylabel('Number of alignments')\n",
    "plt.xlabel('Alignment CIGAR quality')\n",
    "plt.savefig('Bacteria_distribution_CIGAR_quality.png')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e654c590",
   "metadata": {},
   "outputs": [],
   "source": [
    "n, bins, patches = plt.hist(x=qlength, bins='auto', color='r',\n",
    "                            alpha=0.9, rwidth=0.85)\n",
    "plt.grid(axis='y', alpha=0.75)\n",
    "plt.xlabel('Alignment Length')\n",
    "plt.ylabel('Number of alignments')\n",
    "plt.title('Bacteria lf in range 1000')\n",
    "maxfreq = n.max()\n",
    "# Set a clean upper y-axis limit.\n",
    "plt.ylim(ymax=np.ceil(maxfreq / 10) * 10 if maxfreq % 10 else maxfreq + 10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d5232a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "file = pysam.AlignmentFile(\"new_map_sorted.bam\",\"rb\")\n",
    "cq = 1\n",
    "soft_dict = {}\n",
    "for read in file.fetch():\n",
    "    if cq < 4000000:\n",
    "        if prim[read.qname] == AlignInfo(True,False):\n",
    "            # Do analyze here\n",
    "            Cigar_qual, qleng, soft, hard = cigar_compute(read)\n",
    "            bacteria = read.reference_name[0:2]\n",
    "            if bacteria not in soft_dict:\n",
    "                soft_dict[bacteria] = [soft]\n",
    "            else:\n",
    "                soft_dict[bacteria] += [soft]\n",
    "\n",
    "        if cq%100000 == 0:\n",
    "            print('Pass 100K')\n",
    "    else:\n",
    "        break\n",
    "    cq+=1\n",
    "\n",
    "plt.savefig('Bacteria_distribution.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ff0ce2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "cq = 1\n",
    "map_qual = {}\n",
    "check = 0\n",
    "for read in file.fetch():\n",
    "    if cq < 4000000:\n",
    "        if read.is_unmapped or read.is_supplementary or read.is_secondary:\n",
    "            continue\n",
    "        else:\n",
    "            # It is primary alignment\n",
    "            # Do analyze here\n",
    "            mqual = read.mapping_quality\n",
    "            if mqual != 60:\n",
    "                check += 1\n",
    "            bacteria = read.reference_name[0:2]\n",
    "            if bacteria not in map_qual:\n",
    "                map_qual[bacteria] = [read.mapping_quality]\n",
    "            else:\n",
    "                map_qual[bacteria] += [read.mapping_quality]\n",
    "\n",
    "        if cq%100000 == 0:\n",
    "            print('Pass 100K')\n",
    "    else:\n",
    "        break\n",
    "    cq+=1\n",
    "\n",
    "plt.savefig('Bacteria_distribution.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6157554",
   "metadata": {},
   "outputs": [],
   "source": [
    "bac_key = ['lf','lm','cn','ef','ec','pa','bs','sc','sa','se']\n",
    "fig, axi = plt.subplots(nrows=10, ncols=1, figsize = (24,20), sharey = True)\n",
    "for i in range(len(bac_key)):\n",
    "    # Add subplots\n",
    "    axi[i].hist(x=map_qual[bac_key[i]], bins='auto', color='r',alpha=0.9, rwidth=0.85)\n",
    "    # Show texts\n",
    "    axi[i].text(0.2, 0.5, bac_key[i], horizontalalignment='center', verticalalignment='center', transform=axi[i].transAxes)\n",
    "    \n",
    "plt.ylabel('Number of alignments')\n",
    "plt.xlabel('Mapping quality')\n",
    "plt.savefig('Bacteria_distribution_mapqual.png')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c84cdbcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "file = pysam.AlignmentFile(\"new_map_sorted.bam\",\"rb\")\n",
    "cq = 1\n",
    "num_read = 0\n",
    "truth_distri = {}\n",
    "\n",
    "for read in file.fetch():\n",
    "    if cq < 4000000:\n",
    "        if read.is_unmapped or read.is_supplementary or read.is_secondary:\n",
    "            # Discard unmapped, secondary or supplementary\n",
    "            continue\n",
    "        else:\n",
    "            # It is primary alignment\n",
    "            # Do analyze here\n",
    "            # Discard lf, sc, and cn\n",
    "            bacteria = read.reference_name[0:2]\n",
    "            if bacteria == 'lf' or bacteria == 'sc' or bacteria=='cn':\n",
    "                continue\n",
    "            else:\n",
    "                # Check alignment length\n",
    "                qleng = cigar_compute(read)[1]\n",
    "                if qleng < 3000 or qleng > 7000:\n",
    "                    continue\n",
    "                else:\n",
    "                    # CIGAR quality\n",
    "                    align_qual = cigar_compute(read)[0]\n",
    "                    if align_qual < 0.94:\n",
    "                        continue\n",
    "                    else:\n",
    "                        if read.mapping_quality != 60:\n",
    "                            continue\n",
    "                        else:\n",
    "                            # Finish all filter\n",
    "                            num_read += 1\n",
    "                            if bacteria not in truth_distri:\n",
    "                                truth_distri[bacteria] = 1\n",
    "                            else:\n",
    "                                truth_distri[bacteria] += 1\n",
    "        if cq%100000 == 0:\n",
    "            print('Pass 100K')\n",
    "    else:\n",
    "        break\n",
    "    cq+=1\n",
    "\n",
    "plt.savefig('Bacteria_distribution.png')\n",
    "print(num_read)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "585e047c",
   "metadata": {},
   "source": [
    "# After restricting some parameters, we output the data to txt file, which will be used for the model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11370798",
   "metadata": {},
   "outputs": [],
   "source": [
    "@dataclass\n",
    "class AlignInfo:\n",
    "    \"\"\"Class for keeping track of primary and other if available.\"\"\"\n",
    "    primary: bool\n",
    "    other: bool\n",
    "        \n",
    "@dataclass\n",
    "class Overlap:\n",
    "    current: str\n",
    "    after : str"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a65b0a5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "file = pysam.AlignmentFile(\"new_map_sorted.bam\",\"rb\")\n",
    "\n",
    "prim = defaultdict(lambda: AlignInfo(False,False))\n",
    "\n",
    "for read in file.fetch():\n",
    "    qname = read.qname\n",
    "    if read.is_unmapped == False and read.is_supplementary == False and read.is_secondary == False:\n",
    "        prim[qname].primary = True\n",
    "    else:\n",
    "        prim[qname].other = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bc25133",
   "metadata": {},
   "outputs": [],
   "source": [
    "bac_key = ['lm','ef','ec','pa','bs','sa','se']\n",
    "num_read = 0\n",
    "check = 0\n",
    "name_ref = []\n",
    "for read in file.fetch():\n",
    "    if check < 4000000:\n",
    "        if prim[read.qname] == AlignInfo(True, False):\n",
    "            bacteria = read.reference_name[0:2]\n",
    "            if bacteria in bac_key:\n",
    "                #print(qname, bacteria)\n",
    "                align_qual, qleng, soft, hard = cigar_compute(read)\n",
    "                if qleng >= 3000 and qleng <=7000:\n",
    "                    if align_qual >= 0.94:\n",
    "                        if read.mapping_quality == 60:\n",
    "                            num_read += 1\n",
    "                            name_ref += [[read.qname,read.rname, read.reference_start, read.reference_end,read.query_alignment_sequence,bacteria]]\n",
    "    else:\n",
    "        break\n",
    "    check += 1\n",
    "print(num_read)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3cf2c18a",
   "metadata": {},
   "outputs": [],
   "source": [
    "name_ref_sorted = sorted(name_ref, key=lambda tup: (tup[1],tup[2],tup[3]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ea2b9f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def percen(l1,l2):\n",
    "    #both of length 4, but we focus on the last 2 coordinates\n",
    "    #Sorted, so same rname and l1[2] < l2[2]\n",
    "    #Already check intersection. So no need to check here\n",
    "    if l2[3] < l1[3]:\n",
    "        # l1 is longer. Overlap is the whole l2\n",
    "        return (l2[3]-l2[2])/(l1[3]-l1[2])\n",
    "    else:\n",
    "        # Overlap is l1[3]-l2[2]\n",
    "        # Need to compare length\n",
    "        if l2[3]-l2[2] < l1[3]-l1[2]: # list2 < list1\n",
    "            return (l1[3]-l2[2])/(l1[3]-l1[2])\n",
    "        else:\n",
    "            return (l1[3]-l2[2])/(l2[3]-l2[2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2e202ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "ran_pick = random.sample([i for i in range(670951)], 60000)\n",
    "with open('train60k.txt', 'w') as training:\n",
    "    for pick in ran_pick:\n",
    "        see_list = see[pick].split('\\t')\n",
    "        seq = see_list[1]\n",
    "        bac = see_list[2]\n",
    "        over = see_list[3]\n",
    "        training.write(str(pick)+ '/t' + seq + '/t' + bac + '/t' + over)\n",
    "        training.write('\\n')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
